### PR TITLE
DataFrame owned SessionState

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #4617

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/dataframe.rs [hunk 1]
```diff
@@ -74,40 +73,22 @@ use crate::prelude::SessionContext;
...
-            // be written to. As for eventual modifications that would be applied to the
-            // original state after it has been cloned, they will not be picked up by the
-            // clone but that is okay, as it is equivalent to postponing the state update
-            // by keeping the lock until the end of the function scope.
-            state.clone()
-        };
-
-        state_cloned.create_physical_plan(&self.plan).await
+    pub async fn create_physical_plan(&self) -> Result<Arc<dyn ExecutionPlan>> {
+        self.session_state.create_physical_plan(&self.plan).await
```
Comment (tustvold) [inline datafusion/core/src/dataframe.rs:91]: In practice the lock meant that we were still cloning `SessionState` fairly frequently, better to just be explicit about it and optimise from there

**Comment by alamb** [comment]
🤔 there appears to be some sort of deadlock in the tests

https://github.com/apache/arrow-datafusion/actions/runs/3702235488/jobs/6272289305
```

test sql::select::select_values_list ... ok
test sql::select::test_prepare_statement ... ok
test sql::select::unprojected_filter ... ok
test sql::select::use_between_expression_in_select_query ... ok
test sql::order::sort_with_lots_of_repetition_values ... ok
test sql::set_variable::set_bool_variable_bad_value ... ok
test sql::set_variable::set_bool_variable has been running for over 60 seconds
test sql::set_variable::set_time_zone has been running for over 60 seconds
Error: The operation was canceled.
```

**Review by alamb — COMMENTED** [review]
Thank you @tustvold  the idea looks great to me, though I think we need to solve the CI deadlock

**Comment by tustvold** [comment]
Lol, even more reason to remove these locks #4617

File: datafusion/core/src/execution/context.rs [hunk 2]
```diff
@@ -363,7 +363,8 @@ impl SessionContext {
             LogicalPlan::SetVariable(SetVariable {
                 variable, value, ..
             }) => {
-                let config_options = &self.state.write().config.config_options;
```
Comment (tustvold) [inline datafusion/core/src/execution/context.rs:366]: I'm actually somewhat surprised this was compiling, there must be some magic going on to extend the lifetime of the temporary lock guard, which is wild

File: datafusion/core/src/dataframe.rs [hunk 3]
```diff
@@ -74,40 +73,28 @@ use crate::prelude::SessionContext;
...
-            // We need to clone `state` to release the lock that is not `Send`. We could
-            // make the lock `Send` by using `tokio::sync::Mutex`, but that would require to
-            // propagate async even to the `LogicalPlan` building methods.
-            // Cloning `state` here is fine as we then pass it as immutable `&state`, which
-            // means that we avoid write consistency issues as the cloned version will not
-            // be written to. As for eventual modifications that would be applied to the
-            // original state after it has been cloned, they will not be picked up by the
-            // clone but that is okay, as it is equivalent to postponing the state update
-            // by keeping the lock until the end of the function scope.
-            state.clone()
```
Comment (tustvold) [inline datafusion/core/src/dataframe.rs:107]: In practice the lock meant that we were still cloning SessionState fairly frequently, better to just be explicit about it and optimise from there

File: datafusion/core/src/dataframe.rs [hunk 4]
```diff
@@ -74,40 +73,28 @@ use crate::prelude::SessionContext;
...
-        };
+    pub async fn create_physical_plan(mut self) -> Result<Arc<dyn ExecutionPlan>> {
+        self.create_physical_plan_impl().await
+    }
 
-        state_cloned.create_physical_plan(&self.plan).await
+    /// Temporary pending #4626
```
Comment (tustvold) [inline datafusion/core/src/dataframe.rs:94]: This is temporary pending moving the physical plan lowering onto DataFrame from SessionState which is tracked by #4626 and #4629

File: datafusion/core/src/dataframe.rs [hunk 5]
```diff
@@ -74,40 +73,28 @@ use crate::prelude::SessionContext;
...
 pub struct DataFrame {
-    session_state: Arc<RwLock<SessionState>>,
+    session_state: SessionState,
     plan: LogicalPlan,
 }
 
 impl DataFrame {
     /// Create a new Table based on an existing logical plan
-    pub fn new(session_state: Arc<RwLock<SessionState>>, plan: LogicalPlan) -> Self {
+    pub fn new(session_state: SessionState, plan: LogicalPlan) -> Self {
```
Comment (alamb) [inline datafusion/core/src/dataframe.rs:82]: ❤️

File: datafusion/core/src/dataframe.rs [hunk 6]
```diff
@@ -620,28 +604,25 @@ impl DataFrame {
     }
 
     /// Write a `DataFrame` to a CSV file.
-    pub async fn write_csv(self, path: &str) -> Result<()> {
-        let state = self.session_state.read().clone();
```
Comment (alamb) [inline datafusion/core/src/dataframe.rs:624]: yeah all the explicit cloning is definitely a warning sign

**Review by alamb — APPROVED** [review]
I like it -- thank you @tustvold 

There is likely a change of behavior here:

Specifically prior to this PR if you made a dataframe and then modified the SessionContext the dataframe might see some of those changes. 

However I think the old behavior would be quite surprising and so I like this change a lot

**Comment by mingmwang** [comment]
@tustvold @alamb 

I'm a little surprised by this change. In my opinion,  `SessionState` should be the internal state of a `SessionContext`, it can be borrowed/shared, making `SessionState` owned by the `DataFrame` will break the purpose that it is the state of the session. We should not have many `SessionState`s for a single Session in the system.

**Comment by tustvold** [comment]
The idea is that DataFrame contains a snapshot of SessionContext, i.e. a clone of SessionState?

**Comment by alamb** [comment]
@mingmwang  I think the key proposal by @tustvold  is that once a query is planned, it has only a read-only copy of the `SessionState` rather than mutable access to the current SessionState

This has several seemingly nice properties, such as a configuration value changes that are made *after* a query is planned, will not affect the plan (it will only affect subsequently planned queries). Do you agree this sounds like reasonable behvior?

Do you know of any usecases where running a query needs to modify the session context?

**Pull request merged**